core/arch/arm/mm/core_mmu.c

```diff
@@ -367,29 +367,43 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
 
 	memory_map[last].type = MEM_AREA_NOTYPE;
 
+
+
+	/*
+	 * Assign region sizes, note that MEM_AREA_TEE_RAM always uses
+	 * SMALL_PAGE_SIZE if paging is enabled.
+	 */
+	for (map = memory_map; map->type != MEM_AREA_NOTYPE; map++) {
+		paddr_t mask = map->pa | map->size;
+
+		if (!(mask & CORE_MMU_PGDIR_MASK))
+			map->region_size = CORE_MMU_PGDIR_SIZE;
+		else if (!(mask & SMALL_PAGE_MASK))
+			map->region_size = SMALL_PAGE_SIZE;
+		else
+			panic("Impossible memory alignment");
+	}
+
 	/*
 	 * bootcfg_memory_map is sorted in order first by type and last by
 	 * address. This puts TEE_RAM first and TA_RAM second
 	 *
 	 */
-
 	map = memory_map;
 	assert(map->type == MEM_AREA_TEE_RAM);
 	map->va = map->pa;
 #ifdef CFG_WITH_PAGER
```

Comment:
I'm not too happy about this `region_size`, it's used in `core_mmu_v7.c` to select level 2 mapping instead of level 1 mapping. When paging is enabled the area defined by `MEM_AREA_TEE_RAM` has to be mapped 4k pages.
Long term I'd like have the translation tables initialized automatically as for the LPAE case. To map `MEM_AREA_TEE_RAM` using 4k pages we could call `core_mmu_divide_block()` when initializing paging.